app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php

```diff
@@ -276,4 +283,83 @@ private function isCategoryRewritesEnabled()
     {
         return (bool)$this->config->getValue('catalog/seo/generate_category_product_rewrites');
     }
+
+    /**
+     * Return product changed categories store root category ids
+     *
+     * @param Product $product
+     * @return int[]
+     */
+    private function getChangedCategoriesStoreIds(Product $product): array
+    {
+        $oldCategoryIds = $product->getOrigData('category_ids');
+        $actualCategoryIds = $product->getCategoryIds();
+        $result = [];
+        if ($oldCategoryIds === null) {
+            return $result;
+        }
```

Comment:
As $product->getCategoryIds() may cost us extra request to Db it's better to end up earlier.
Also, why we skip url rewrite generation in case product didn't have categories but currently maybe does?
```suggestion
        if ($oldCategoryIds === null) {
            return $result;
        }
        $actualCategoryIds = $product->getCategoryIds();
        $result = [];
       
```